certora/specs/Pausable.spec

```diff
@@ -0,0 +1,93 @@
+import "helpers.spec"
+import "methods/IPausable.spec"
+
+methods {
+    pause()
+    unpause()
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: _pause pauses the contract                                                                    │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pause(env e) {
+    require nonpayable(e);
+
+    bool pausedBefore = paused();
+
+    pause@withrevert(e);
+    bool success = !lastReverted;
+
+    bool pausedAfter = paused();
+    
+    // liveness
+    assert success => !pausedBefore, "unpause call must succeed if the contract was unpaused before";

```

Comment:
liveness property should be equivalence `<=>`. 

Right now you are proving that is the call was succesfull, it wasn't paused before, but you are not proving that is the contract wasn't paused before, then the call doesn't revert

Reply:
Got it, I didn't make it work at first and didn't try again. Now it does!

Is this "liveness property should be equivalence `<=>`" a common rule or something that just applies here?

Reply:
The issue is that sometimes the operation can fail for "conditions" that are hard/impossible to formalize. For example, in a TimelockController, the execution will revert if the timelocked call (which is an arbitrary call) reverts. In that case its very hard to prove lifeness, and we are limited to proving the first point.

Reply:
Thanks for confirming, so far I've seen `success <=> ...` for liveness across other examples in this repository, but wasn't clear to me if that was a general rule.

I'll use it moving forward with the DoubleEndedQueue.

> the execution will revert if the timelocked call (which is an arbitrary call) reverts

Isn't this most of the reasons is difficult to fully formally verify the library? Now I see why generalization gets hard in this context.